server/services/v1/database/query_index.go

```diff
@@ -0,0 +1,294 @@
+
+	it, err := tx.ReadRange(ctx, start, end, false)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &IndexIterator{
+		tx:   tx,
+		it:   it,
+		keys: []keys.Key{start, end},
+	}, nil
+}
+
+func (q *QueryIndex) IndexInfo(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection) (*QueryIndexInfo, error) {
+	rowCountKey := genRowCountKey(coll)
+	indexSizeKey := genIndexSizeKey(coll)
+
+	rows, err := tx.AtomicRead(ctx, rowCountKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	size, err := tx.AtomicRead(ctx, indexSizeKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &QueryIndexInfo{
+		rows,
+		size,
+	}, nil
+}
+
+func (q *QueryIndex) Index(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, doc []byte, id []interface{}) error {
+
+	indexKeys, size, rowCount := q.buildKVs(coll, doc, id)
+
+	rowCountKey := genRowCountKey(coll)
+	indexSizeKey := genIndexSizeKey(coll)
+
+	tx.AtomicAdd(ctx, rowCountKey, rowCount)
+	tx.AtomicAdd(ctx, indexSizeKey, size)
+
+	for _, indexKey := range indexKeys {
+		err := tx.Insert(ctx, indexKey, nil)
+
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func (q *QueryIndex) buildKVs(coll *schema.DefaultCollection, doc []byte, id []interface{}) (indexedKeys []keys.Key, sizeIncrease int64, rowCount int64) {
+
+	for _, field := range coll.QueryableFields {
+
+		if schema.IsReservedField(field.Name()) {
+			continue
+		}
+
+		var rows []FieldIndex
+		if field.DataType == schema.ArrayType {
+			rows = indexArray(doc, field, field.KeyPath())
+		} else {
+			row, err := indexField(doc, field.FieldName, field.DataType, 0, field.KeyPath()...)
+
+			if err != nil {
+				log.Err(err).Msgf("Failed to index field name: %s", field.FieldName)
+				continue
+			}
+
+			rows = []FieldIndex{*row}
+		}
+
+		for _, row := range rows {
+			indexKey := newKeyWithId(id, coll.EncodedName, coll.Indexes.QueryIndex.Name, "kvs", row.name, coll.IdxVer, row.value.AsInterface(), row.pos)
+
+			indexedKeys = append(indexedKeys, indexKey)
+			rowCount += 1
+			sizeIncrease += int64(len(indexKey.SerializeToBytes()))
+		}
+	}
+
+	return
+}
+
+func indexField(doc []byte, fieldName string, dataType schema.FieldType, pos int, keyPath ...string) (*FieldIndex, error) {
+	val, _, _, err := jsonparser.Get(doc, keyPath...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	row, err := newFieldIndex(dataType, fieldName, val, pos)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return row, nil
+}
+
+func IndexNestedField(doc []byte, topField string, pos int) []FieldIndex {
+	var indexedFields []FieldIndex
+	processor := func(key []byte, value []byte, jsonDataType jsonparser.ValueType, offset int) error {
+		fieldType := schema.ToFieldType(jsonDataType.String(), "", "")
+
+		switch fieldType {
+		case schema.ArrayType:
+			// Do not index into nested arrays
+			return nil
```

Comment:
Should we add some dummy for the key so that we know this nested key exists? Then we can use the index to query?